votor/src/consensus_pool/vote_pool.rs

```diff
@@ -2,239 +2,375 @@ use {
+    Duplicate,
+    #[error("slashable behavior")]
+    Slash,
 }
 
+/// Container to store per slot votes.
 #[derive(Default)]
-pub(super) struct SimpleVotePool {
-    votes: Vec<VoteMessage>,
-    total_stake: Stake,
-    prev_voted_validators: BTreeSet<Pubkey>,
+struct Votes {
+    /// Skip votes are stored in map indexed by validator.
+    skip: BTreeMap<Pubkey, VoteMessage>,
+    /// Skip fallback votes are stored in map indexed by validator.
+    skip_fallback: BTreeMap<Pubkey, VoteMessage>,
+    /// Finalize votes are stored in map indexed by validator.
+    finalize: BTreeMap<Pubkey, VoteMessage>,
+    /// Notar votes are stored in map indexed by validator.
+    notar: BTreeMap<Pubkey, VoteMessage>,
+    /// A validator can vote notar fallback on multiple blocks.
+    ///
+    /// Per validator, we store a map of which block ids the validator has voted notar fallback on.
+    notar_fallback: BTreeMap<Pubkey, BTreeMap<Hash, VoteMessage>>,
 }
 
-impl SimpleVotePool {
-    pub(super) fn add_vote(
-        &mut self,
-        validator_vote_key: Pubkey,
-        validator_stake: Stake,
-        vote: VoteMessage,
-    ) -> Option<Stake> {
-        if !self.prev_voted_validators.insert(validator_vote_key) {
-            return None;
+impl Votes {
+    /// Adds votes.
+    ///
+    /// Checks for different types of slashable behavior and duplicate votes returning appropriate errors.
+    fn add_vote(&mut self, voter: Pubkey, vote: VoteMessage) -> Result<(), AddVoteError> {
+        match vote.vote {
+            Vote::Notarize(notar) => {
+                if self.skip.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.notar.entry(voter) {
+                    Entry::Occupied(e) => {
+                        if e.get().vote.block_id().unwrap() == &notar.block_id {
+                            Err(AddVoteError::Duplicate)
+                        } else {
+                            Err(AddVoteError::Slash)
+                        }
+                    }
+                    Entry::Vacant(e) => {
+                        e.insert(vote);
+                        Ok(())
+                    }
+                }
+            }
+            Vote::NotarizeFallback(nf) => {
+                if self.finalize.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.notar_fallback.entry(voter) {
+                    Entry::Vacant(e) => {
+                        let mut map = BTreeMap::new();
+                        map.insert(nf.block_id, vote);
+                        e.insert(map);
+                        Ok(())
+                    }
+                    Entry::Occupied(mut e) => {
+                        let map = e.get_mut();
+                        match map.entry(nf.block_id) {
+                            Entry::Vacant(map_e) => {
+                                map_e.insert(vote);
```

Comment:
if the length would become more than 3 here, that's also a slashable offence

Reply:
good catch, fixed.